tests/unit/test_mongodb_backups.py

```diff
@@ -0,0 +1,225 @@
+
+    @patch("charm.subprocess.check_output")
+    def test_verify_resync_cred_error(self, check_output):
+        """Test _verify_resync goes into blocked state when credential error occurs."""
+        check_output.side_effect = CalledProcessError(
+            cmd="percona-backup-mongodb status", returncode=403
+        )
+        self.harness.charm.backups._verify_resync()
+        self.assertTrue(isinstance(self.harness.charm.unit.status, BlockedStatus))
+
+    @patch("charm.subprocess.check_output")
+    def test_verify_resync_syncing(self, check_output):
+        """Test _verify_resync goes into waiting state when syncing and raises an error."""
+        check_output.return_value = b"Currently running:\n====\nResync op"
+
+        # disable retry
+        self.harness.charm.backups._get_pbm_status.retry.retry = tenacity.retry_if_not_result(
+            lambda x: True
+        )
+
+        with self.assertRaises(ResyncError):
+            self.harness.charm.backups._verify_resync()
+
+        self.assertTrue(isinstance(self.harness.charm.unit.status, WaitingStatus))
+
+    def test_resync_config_options_failure(self):
+        """Verifies _resync_config_options goes into blocked state when it can't start pbm."""
+        mock_snap = mock.Mock()
+        mock_snap.start.side_effect = snap.SnapError
+        self.harness.charm.backups._resync_config_options(mock_snap)
+        self.assertTrue(isinstance(self.harness.charm.unit.status, BlockedStatus))
+
+    @patch("charm.subprocess.check_output")
+    def test_set_config_options(self, check_output):
+        """Verifies _set_config_options failure raises SetPBMConfigError."""
+        check_output.side_effect = (
+            CalledProcessError(
+                cmd="percona-backup-mongodb config --set this_key=doesnt_exist", returncode=42
+            ),
+        )
+        with self.assertRaises(SetPBMConfigError):
+            self.harness.charm.backups._set_config_options({"this_key": "doesnt_exist"})
+
+    def test_backup_without_rel(self):
+        """Verifies no backups are attempted without s3 relation."""
+        action_event = mock.Mock()
+        action_event.params = {}
+
+        self.harness.charm.backups._on_create_backup_action(action_event)
+        action_event.fail.assert_called()
+
+    @patch("charm.subprocess.check_output")
+    @patch("charm.MongoDBBackups._verify_resync")
+    def test_backup_syncing(self, resync, subprocess):
+        """Verifies backup is deferred if more time is needed to resync."""
+        action_event = mock.Mock()
+        action_event.params = {}
+        resync.side_effect = ResyncError
+
+        self.harness.add_relation(RELATION_NAME, "s3-integrator")
+        self.harness.charm.backups._on_create_backup_action(action_event)
+
+        action_event.defer.assert_called()
+
+    @patch("charm.subprocess.check_output")
+    @patch("charm.MongoDBBackups._verify_resync")
+    def test_backup_failed(self, resync, subprocess):
```

Comment:
Is it possible to have a test that checks for a correct backup start? (by checking that `set_results` was called and/or `MaintenanceStatus` was set)

Reply:
Those will be tested in the integration tests :)